### ISSUE TITLE
can we train by Parallel Computing or Multithreading or multi-Progress

### ISSUE THREAD

**Issue by joytianya** [google/sentencepiece#366]
can we train by Parallel Computing or Multithreading or multi-Progress?
Speed up training
thank you

**Comment by yutkin** [comment]
@joytianya Yes, we can! For example, look at [YouTokenToMe](https://github.com/vkcom/youtokentome). This BPE implementation quite efficiently uses parallel processing.

**Comment by taku910** [comment]
Thank you. I will take a look. Actually, the current BPE algorithm is a little conservative to find the most frequent pairs.

**Comment by taku910** [comment]
Will work on it in the next release.

**Comment by heyaudace** [comment]
> Will work on it in the next release.

Hi @taku910 - I was wondering whether this feature was released. Thank you

**Comment by ganeshkrishnan1** [comment]
Just tagging along: is it possible to use multi threaded tokenization for multi-cpu training?

**Comment by DmitriiP20** [comment]
+1 this is badly needed for large corpora and large vocab sizes.

**Comment by taku910** [comment]
Just a heads-up:  We made the BPE training 20x faster on the head, even though this isn't due to parallel processing. We’ll be highlighting this performance gain in the v0.2.2 release notes.

**Comment by taku910** [comment]
We investigated parallelizing BPE training in SentencePiece, prototyping a Hugging Face (HF) style parallel merge approach and analyzing YouTokenToMe's (YTTM) statically partitioned design. 
Our benchmarks on a 19.3MB corpus (~180k unique words, vocab 32,000) show that **parallelizing BPE training does not provide a meaningful speedup and can cause slowdowns:**
*   **1T Baseline**: 6.86s
*   **4T (Parallel Threshold 1000)**: 6.30s (~1.09x speedup)*
*   **4T (Parallel Threshold 10)**: 8.01s (~0.86x speedup - slowdown)
*\*Note: The ~9% speedup is due to parallel sentence normalization during initialization (already in SentencePiece), not the BPE loop itself. The BPE parallelization contribution was <1%. "Threshold" is the minimum pair frequency required to trigger parallel execution.*
#### Key Findings
1.  **Amdahl's Law Constraint**: The parallelizable merge step only accounts for **~24%** of BPE training time in SentencePiece. The remaining **76%** (priority queue updates) is inherently sequential, limiting the maximum theoretical speedup to $\approx 1.3\text{x}$.
2.  **Thread Overhead**: Thread scheduling and synchronization overhead (~160 us per merge) outweighs parallel gains for the long tail of rare merges, causing slowdowns at lower thresholds.
3.  **HF & YTTM Parallel Limits**: 
    *   **HF-style**: Replicates HF's `Rayon` strategy of parallelizing a single pair's merge. It shares the same sequential PQ bottleneck.
    *   **YTTM-style & Scalability**: YTTM achieves sub-linear scaling because its BPE-specific monolithic architecture (compact arrays, custom hash maps, split queues) minimizes the sequential fraction ($S$) of the loop. Because SentencePiece relies on heavier generic structures, its sequential fraction remains too high (~76%) to benefit from YTTM's parallel design without a complete rewrite of the core BPE structures.

**Comment by taku910** [comment]
If there is no update, I will  close this.
